README.md
#  OnBoardingIntro

[![Say Thanks!](https://img.shields.io/badge/Say%20Thanks-!-1EAEDB.svg)](https://saythanks.io/to/xsahil03x) 

A OnBoarding Intro Screen Library for Easy and Quick Usage.
* Checkout the Original Author : [![GitHub followers])](https://github.com/ahmedelsayed96)


The source code is **100% Dart**, and everything resides in the [/lib](https://github.com/pharous-flutter/onboard_intro/tree/master/lib) folder.


### Show some :heart: and star the repo to support the project



[![Open Source Love](https://badges.frapsoft.com/os/v1/open-source.svg?v=102)](https://opensource.org/licenses/MIT)
[![License](https://img.shields.io/badge/license-MIT-blue.svg)](https://github.com/pharous-flutter/onboard_intro/tree/master/LICENSE)


### GIF
<img src="https://github.com/pharous-flutter/onboard_intro/raw/master/assets/onBoarding%20intro.gif" height="400" alt="GIF"/>

## 💻 Installation
In the `dependencies:` section of your `pubspec.yaml`, add the following line:



```
onboarding_intro: <latest_version>
```

## ❔ Usage

### Import this class
```dart
import 'package:onboarding_intro/onboarding_intro.dart';
```

### Create a List of OnBoardingData

```dart
  List<OnBoardingData> data = [
    OnBoardingData('images/onboarding1.svg',
        'First Screen Description...',
        isSvg: true),
    OnBoardingData(
      'images/onboarding2.svg',
      'Second Screen Description...',
      isSvg: true,
    ),
    OnBoardingData(
      'images/onboarding3.svg',
      'Third Screen Description...',
      isSvg: true,
    ),
  ];
```

### Pass it into the OnBoardingIntro widget
```dart
  @override
  Widget build(BuildContext context) {
   return Scaffold(
      key: _key,
      body: OnBoardingIntro(
        isSvgImage: true,
        selectedColor: Color(0xff35200b),
        unSelectedColor: Colors.grey,
        finishText: 'Home',
        nextText: 'Next',
        backText: 'Back',
        skipText: 'Skip',
        onFinalClicked: () {
         Scaffold.of(context).showSnackBar(SnackBar(content: Text('Finished')));
        },
        onSkipClicked: () {
          Scaffold.of(context).showSnackBar(SnackBar(content: Text('Skiped')));
        },
        logoImage: 'images/logo.svg',
        style: TextStyle(
            fontSize: 17, fontWeight: FontWeight.bold, wordSpacing: 2),
        data: data,
      ),
    );
  }
```

## 🎨 Customization and Attributes

#### OnBoardingIntro attributes
<table>
    <th>Attribute Name</th>
    <th>Example Value</th>
    <th>Description</th>
    <tr>
        <td>data</td>
        <td>List&lt;OnBoardingData&gt;</td>
        <td>The list of pages to be displayed</td>
    </tr>
    <tr>
        <td>onFinalClicked</td>
        <td>(){}</td>
        <td>Function to be called on pressing done button</td>
    </tr>
    <tr>
        <td>onSkipClicked</td>
        <td>(){}</td>
        <td>Function to be called on pressing skip button</td>
    </tr>
    <tr>
        <td>finishText</td>
        <td>"Let's Go"</td>
        <td>Done button text "</td>
    </tr>
    <tr>
        <td>skipText</td>
        <td>"Skip"</td>
        <td>Skip button text "</td>
    </tr>
   <tr>
        <td>nextText</td>
        <td>"Next"</td>
        <td>Next button text "</td>
    </tr>
    <tr>
        <td>isSvgImage</td>
        <td>true</td>
        <td>Logo image should be svg or not. Defaults to true</td>
    </tr>
    <tr>
        <td>logoImage</td>
        <td>8.0</td>
        <td>App logo Image</td>
    </tr>
    <tr>
        <td>style</td>
        <td>TextStyle(fontSize: 17, fontWeight: FontWeight.bold, wordSpacing: 2)</td>
        <td>Custom text style for details text</td>
    </tr>
    <tr>
        <td>selectedColor</td>
        <td>Color(0xff35200b)</td>
        <td>selected color  content defualt is Colors.blue</td>
    </tr>
   <tr>
        <td>unSelectedColor</td>
        <td>Color(0xff35200b)</td>
        <td>selected color  content defualt is Colors.grey</td>
    </tr>
</table>

## Contributors ✨

Thanks goes to these wonderful people ([emoji key](https://allcontributors.org/docs/en/emoji-key)):

<table>
  <tr>
   <td align="center"><a href="https://github.com/ahmedelsayed96"><img src="https://avatars1.githubusercontent.com/u/18017854?s=100" width="100px;" alt=""/><br /><sub><b>Ahmed Eslayed</b></sub></a><br /><a href="https://github.com/ahmedelsayed96" title="Coding">💻</a></td>
</table>

# 📃 License

    Copyright (c) 2012 Pharous
    
    Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
    
    The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
    
    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

## Getting Started

For help getting started with Flutter, view our online [documentation](https://flutter.io/).

For help on editing package code, view the [documentation](https://flutter.io/developing-packages/).


CHANGELOG.md
## [0.0.1] - First Release
## [0.0.2] - Example Added
## [0.0.3] - Fix Documentation



test/onboarding_intro_test.dart
// import 'package:flutter_test/flutter_test.dart';
//
// import 'package:onboarding_intro/onboarding_intro.dart';
//
// void main() {
//   test('adds one to input values', () {
//     final calculator = Calculator();
//     expect(calculator.addOne(2), 3);
//     expect(calculator.addOne(-7), -6);
//     expect(calculator.addOne(0), 1);
//     expect(() => calculator.addOne(null), throwsNoSuchMethodError);
//   });
// }


lib/indicator.dart
import 'package:flutter/material.dart';

class DotIndicator extends StatelessWidget {
  int length;
  int currentIndex;
  Color unSelectedColor;
  Color selectedColor;

  DotIndicator(
      {this.currentIndex,
      this.length = 0,
      this.selectedColor = Colors.blue,
      this.unSelectedColor = Colors.grey});

  @override
  Widget build(BuildContext context) {
    return buildIndicator(context);
  }

  Widget buildIndicator(BuildContext context) {
    return Container(
      child: Center(
        child: Row(
          mainAxisAlignment: MainAxisAlignment.center,
          children: List.generate(
            length,
            (index) => buildDot(index, context),
          ),
        ),
      ),
    );
  }

  Widget buildDot(int index, BuildContext context) {
    return AnimatedContainer(
      duration: Duration(milliseconds: 300),
      height: 8,
      margin: EdgeInsets.symmetric(horizontal: 1),
      width: currentIndex == index ? 25 : 10,
      decoration: BoxDecoration(
        borderRadius: BorderRadius.circular(20),
        color: currentIndex == index ? selectedColor : unSelectedColor,
      ),
    );
  }
}


lib/onboarding_intro.dart
library onboarding_intro;

import 'package:flutter/material.dart';

import 'image.dart';
import 'indicator.dart';
import 'on_boarding_model.dart';

class OnBoardingIntro extends StatefulWidget {
  final List<OnBoardingData> data;
  final String logoImage;
  final bool isSvgImage;
  final Function onFinalClicked;
  final Function onSkipClicked;
  final TextStyle style;
  final Color selectedColor;
  final Color unSelectedColor;
  final String finishText;
  final String skipText;
  final String nextText;
  final String backText;

  OnBoardingIntro(
      {@required this.data,
      this.isSvgImage = false,
      this.logoImage,
      this.selectedColor=Colors.blue,
      this.onSkipClicked,
      this.unSelectedColor=Colors.grey,
      this.onFinalClicked,
      this.finishText,
      this.skipText,
      this.nextText,
      this.backText,
      this.style});

  @override
  _OnBoardingIntroState createState() => _OnBoardingIntroState();
}

class _OnBoardingIntroState extends State<OnBoardingIntro> {
  List<OnBoardingData> data;

  // OnBoardingData onBoardingData;
  int currentIndex = 0;
  PageController controller = PageController();

  @override
  void initState() {
    data = widget.data;
    // onBoardingData = data.first;
    controller.addListener(() {
      if (currentIndex == controller.page.toInt()) return;
      setState(() {
        currentIndex = controller.page.round();
        print('inedx : $currentIndex');
      });
    });
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return SafeArea(
      child: Padding(
        padding: const EdgeInsets.all(16.0),
        child: Column(
          children: [
            Expanded(
              child: PageView(
                controller: controller,
                children: data.map((e) => buildBody(e)).toList(),
              ),
            ),
            SizedBox(
              height: 0,
            ),
            DotIndicator(
              selectedColor: widget.selectedColor,
              unSelectedColor: widget.unSelectedColor,
              length: data.length,
              currentIndex: currentIndex,
            ),
            SizedBox(
              height: 16,
            ),
            buildButtons(),
            SizedBox(
              height: 16,
            ),
          ],
        ),
      ),
    );
  }

  Widget buildBody(OnBoardingData onBoardingData) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      mainAxisAlignment: MainAxisAlignment.spaceAround,
      children: [
        IntroImage(
          isSvgImage: widget.isSvgImage,
          image: widget.logoImage,
        ),
        buildImages(onBoardingData),

        Text(
          onBoardingData.description,
          textAlign: TextAlign.center,
          style: widget.style,
        ),

        // PageView(
        //   controller: controller,
        //   children: data.map((e) => buildImages(e)).toList(),
        // ),
      ],
    );
  }

  Row buildButtons() {
    bool isLast = (currentIndex == data.length - 1);
    return Row(
      children: [
        Expanded(
          flex: 2,
          child: buildButton(isLast ? widget.finishText : widget.nextText,
              color: widget.selectedColor,
              borderColor: widget.selectedColor,
              showIcon: isLast ? Icons.done : Icons.arrow_back_ios_outlined,
              onTap: onNextClicked),
        ),
        SizedBox(width: 16),
        Expanded(
            child: buildButton(
                currentIndex == 0 ? widget.skipText : widget.backText,
                onTap: onSkipBackClicked,
                color: Colors.transparent,
                borderColor: Colors.black,
                textColor: Colors.black)),
      ],
    );
  }

  Widget buildButton(String text,
      {Color color,
      Color borderColor,
      IconData showIcon,
      Color textColor = Colors.white,
      Function onTap}) {
    return InkWell(
      onTap: onTap,
      child: AnimatedContainer(
        duration: Duration(milliseconds: 300),
        decoration: BoxDecoration(
            border: Border.all(color: borderColor),
            color: color,
            borderRadius: BorderRadius.circular(5)),
        padding: EdgeInsets.all(8),
        child: Row(
          crossAxisAlignment: CrossAxisAlignment.center,
          children: [
            if (showIcon != null)
              Icon(
                showIcon,
                color: Colors.white,
              ),
            Expanded(
              child: Container(
                height: 24,
                child: Center(
                  child: Text(
                    text,
                    textAlign: TextAlign.center,
                    style:
                        TextStyle(color: textColor, fontWeight: FontWeight.bold),
                  ),
                ),
              ),
            ),
            if (showIcon != null)
              SizedBox(
                width: 24,
              )
          ],
        ),
      ),
    );
  }

  Widget buildImages(OnBoardingData onBoardingData) {
    return IntroImage(
      isSvgImage: onBoardingData.isSvg,
      image: onBoardingData.image,
      width: onBoardingData.imageWidth,
      height: onBoardingData.imageHeight,
    );
  }

  void onNextClicked() {
    if (currentIndex == data.length - 1)
      widget.onFinalClicked();
    else
      setState(() {
        controller.animateToPage(currentIndex + 1,
            duration: Duration(milliseconds: 300), curve: Curves.easeIn);
        // onBoardingData = data[currentIndex];
      });
  }

  void onSkipBackClicked() {
    if (currentIndex == 0 && widget.onSkipClicked != null)
      widget.onSkipClicked();
    else
      setState(() {
        controller.animateToPage(currentIndex - 1,
            duration: Duration(milliseconds: 300), curve: Curves.easeIn);
        // onBoardingData = data[currentIndex];
      });
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:onboarding_intro/on_boarding_model.dart';
import 'package:onboarding_intro/onboarding_intro.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  List<OnBoardingData> data = [
    OnBoardingData('images/onboarding1.svg',
        'First Screen Description...',
        isSvg: true),
    OnBoardingData(
      'images/onboarding2.svg',
      'Second Screen Description...',
      isSvg: true,
    ),
    OnBoardingData(
      'images/onboarding3.svg',
      'Third Screen Description...',
      isSvg: true,
    ),
  ];

  var _key;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      key: _key,
      body: OnBoardingIntro(
        isSvgImage: true,
        selectedColor: Color(0xff35200b),
        unSelectedColor: Colors.grey,
        finishText: 'Home',
        nextText: 'Next',
        backText: 'Back',
        skipText: 'Skip',
        onFinalClicked: () {
         Scaffold.of(context).showSnackBar(SnackBar(content: Text('Finished')));
        },
        onSkipClicked: () {
          Scaffold.of(context).showSnackBar(SnackBar(content: Text('Skiped')));
        },
        logoImage: 'images/logo.svg',
        style: TextStyle(
            fontSize: 17, fontWeight: FontWeight.bold, wordSpacing: 2),
        data: data,
      ),
    );
  }
}
